mlkem/indcpa.c

```diff
@@ -155,34 +157,69 @@ static unsigned int rej_uniform(int16_t *r,
 #define GEN_MATRIX_NBLOCKS ((12*KYBER_N/8*(1 << 12)/KYBER_Q + XOF_BLOCKBYTES)/XOF_BLOCKBYTES)
 // Not static for benchmarking
 void gen_matrix(polyvec *a, const uint8_t seed[KYBER_SYMBYTES], int transposed) {
-    unsigned int ctr, i, j, k;
-    unsigned int buflen, off;
-    uint8_t buf[GEN_MATRIX_NBLOCKS * XOF_BLOCKBYTES + 2];
+    unsigned int ctr, i, j;
+    unsigned int buflen;
+    uint8_t bufx[KECCAK_WAY][GEN_MATRIX_NBLOCKS *XOF_BLOCKBYTES + 2];
+    uint8_t *buf = NULL;
+    int16_t *vec = NULL;
     xof_state state;
 
-    for (i = 0; i < KYBER_K; i++) {
-        for (j = 0; j < KYBER_K; j++) {
+    keccakx kx;
+    uint8_t seedxy[KECCAK_WAY][KYBER_SYMBYTES + 2];
+
+    for (i = 0; i < (KYBER_K *KYBER_K / KECCAK_WAY) *KECCAK_WAY; i += KECCAK_WAY) {
+        for (j = 0; j < KECCAK_WAY; j++) {
+            memcpy(seedxy[j], seed, KYBER_SYMBYTES);
             if (transposed) {
-                xof_absorb(&state, seed, i, j);
+                seedxy[j][KYBER_SYMBYTES + 0] = (i + j) / KYBER_K;
+                seedxy[j][KYBER_SYMBYTES + 1] = (i + j) % KYBER_K;
             } else {
-                xof_absorb(&state, seed, j, i);
+                seedxy[j][KYBER_SYMBYTES + 0] = (i + j) % KYBER_K;
+                seedxy[j][KYBER_SYMBYTES + 1] = (i + j) / KYBER_K;
             }
 
-            xof_squeezeblocks(buf, GEN_MATRIX_NBLOCKS, &state);
-            buflen = GEN_MATRIX_NBLOCKS * XOF_BLOCKBYTES;
-            ctr = rej_uniform(a[i].vec[j].coeffs, KYBER_N, buf, buflen);
-
-            while (ctr < KYBER_N) {
-                off = buflen % 3;
-                for (k = 0; k < off; k++) {
-                    buf[k] = buf[buflen - off + k];
-                }
-                xof_squeezeblocks(buf + off, 1, &state);
-                buflen = off + XOF_BLOCKBYTES;
-                ctr += rej_uniform(a[i].vec[j].coeffs + ctr, KYBER_N - ctr, buf, buflen);
+            keccakx_set_lane_input(&kx, j, seedxy[j], KYBER_SYMBYTES + 2);
+            keccakx_set_lane_output(&kx, j, bufx[j], GEN_MATRIX_NBLOCKS *XOF_BLOCKBYTES + 2);
+
+        }
+        shake128x_absorb(&kx);
+        shake128x_squeezeblocks(&kx, GEN_MATRIX_NBLOCKS);
+
+        for (j = 0; j < KECCAK_WAY; j++) {
+            vec = a[ (i + j) / KYBER_K].vec[ (i + j) % KYBER_K].coeffs;
+            buf = bufx[j];
+            buflen = GEN_MATRIX_NBLOCKS *XOF_BLOCKBYTES;
+            ctr = rej_uniform(vec, KYBER_N, buf, buflen);
+
+            while (ctr < KYBER_N ) {
+                uint64_t *ctx = keccakx_get_lane_state(&kx, j);
```

Comment:
switch to single keccak here. We can collect all single keccak and write a 2nd pass, it's probably faster. However, it will make the code logic a little complicated.